### DISCUSSION TITLE
Modern Modular Development

### DISCUSSION THREAD

**Discussion by larry-tx** [Laravel-Backpack/community-forum#319] [General]
### Backpack, Laravel, PHP, DB version:
Backpack v. 3.5
Laravel v. 5.7
PHP v. 7.2.10
MySQL v. 8.0.13

I'v worked through the BackPack documentation, and I would, first, like to compliment you on your documentation. Too many developers seem to think that users should be able to intuit how their product works. I am always inclined to avoid products that don't have adequate and clear documentation.

Unfortunately, in studying the Backpack documentation, I was unable to find something that is essential to any modern development project except for the most simplistic, trivial project, and that is the ability to manage the modules within the project. The application that I am working on rebuilding in Laravel has 222 tables (with associated models) grouped into 27 modules with cross-module relationships. (Those models will never, ever be dumped, willy-nilly into the App directory totally unmanaged.) Yii2 was quite capable of handling that development. Having seen the nWidart/laravel-modules package, I was hoping that Laravel had grown beyond a hobbyist platform for personal blogs. Widart's package very nicely, competently and professionally handles modular development.

At this point, I have to ask whether Backpack already has this capability built-in? If not surely this is something that you are diligently working on and near-ready to roll out? My need for this capability is immediate. How soon do you anticipate rolling out the capability?

In the astounding possibility that you aren't quite ready to roll out modular capability, do you have any recommendations on how to incorporate nWidart/laravel-modules into Backpack? I'm not wedded to that specific package (although) it is the best that I've been able to find), so if you're looking at some other approach, I'm willing to consider it.

This is kind of in-line with the fact that Backpack continues to use Bootstrap 3 because of AdminLTE. My hosting provider has banned Bootstrap 3 because of serious vulnerabilities. Long, long ago, when Bootstrap 4 came out, I became so enamored with Grid and Flexbox that I will never go back. Abdullah Almsaeed has clearly decided that AdminLTE is dying and best left to a morgue. By the time they get ready to launch a product based on Bootstrap 4 (if they last that long), Bootstrap 5 will be out and they will remain a purely historical product.

Which makes it unimaginable that Backpack is still doddering around wondering what on earth to do. In case you haven't noticed, this is a software industry, an industry to moves extraordinarily fast and waits on no one. As I said before, Bootstrap 4 came out an extraordinarily long time ago, and the overwhelmingly vast majority of developers have moved to Grid and Flexbox (and Bootstrap 4), also a very long time ago. Is Backpack going to come into the modern age any time soon?

Needless to say, I'm doing to have to come to some decisions very soon. Is Backpack going to stay in an age of trivial applications based on antiquated technology, or is it worth my time to master it and develop in it?

**Comment by welcome** [comment]
Hello there! Thanks for opening your first issue on this repo!

Just a heads-up: **Here at Backpack we use Github Issues only for tracking bugs**. Talk about new features is also acceptable. This helps _a lot_ in keeping our focus on improving Backpack. **If you issue is not a bug/feature, please help us out by closing the issue yourself and posting in the appropriate medium (see below)**. If you're not sure where it fits, it's ok, a community member will probably reply to help you with that.

Backpack communication mediums:
 - Bug Reports, Feature Requests - Github Issues (here);
 - Quick help (_How do I do X_) - [Gitter Chatroom](gitter.im/BackpackForLaravel/Lobby);
 - Long questions (_I have done X and Y and it won't do Z wtf_) - [Stackoverflow](https://stackoverflow.com/questions/tagged/backpack-for-laravel), using the ```backpack-for-laravel``` tag;


Please keep in mind Backpack offers no official / paid support. Whatever help you receive here, on Gitter, Slack or Stackoverflow is thanks to our awesome _awesome_ community members, who give up some of their time to help their peers. If you want to join our community, just start pitching in. We take pride in being a welcoming bunch.

Thank you!

--
Justin Case
The Backpack Robot

**Comment by tabacitu** [comment]
Hi @LarryTX ,

Thanks for the carefully written issue - I appreciate you took the time.

I agree with moving to Bootstrap 4. Backpack v4 (we're working on it right now) will have BS4 by default, and the default theme will use Tabler instead of AdminLTE. I think it's unfortunate they haven't upgraded too, but we've waited long enough, just like you said. Can't wait for flexbox either, I usually use BS4 on front-end and BS3 on back-end, just because of AdminLTE... time to stop :-)

I know of nWidart/laravel-modules and I think they've done a very good job. You should be able to use it with Backpack, but Backpack won't make it easy for you to do so (we don't have any guides or anything). Yes, we are considering using Modular Development for v4, but I can't assure you we will. There are downsides to it too. I guess we can only promise that a final decision will be evident in 1-2 months when Backpack v4 is out. If you like their approach and need modular development ASAP, maybe take a look at AsgardCMS - it uses this package.

Cheers!

**Comment by tabacitu** [comment]
Update: just learned about [InterNACHI/modular](https://github.com/InterNACHI/modular?ref=laravelnews) which I believe is  a lot less opinionated, and easier to use - because it just uses the Laravel conventions. Perhaps we should consider using it to generate a module for each CRUD we create.

I'll move this to Backpack/CRUD so that we can continue the discussion.

**Reply by tabacitu** [reply]
Clarification - I think this should _maybe_ do this as an optional. For example Generators could generate a Module instead of putting the CRUD files inside the default folders.

**Comment by welcome** [comment]
Hello there! Thanks for opening your first issue on this repo!

Just a heads-up: **Here at Backpack we use Github Issues only for tracking bugs**. Talk about new features is also acceptable. This helps _a lot_ in keeping our focus on improving Backpack. **If you issue is not a bug/feature, please help us out by closing the issue yourself and posting in the appropriate medium (see below)**. If you're not sure where it fits, it's ok, a community member will probably reply to help you with that.

Backpack communication channels:
 - Bug Reports, Feature Requests - Github Issues (here);
 - Quick help (_How do I do X_) - [Gitter Chatroom](https://gitter.im/BackpackForLaravel/Lobby);
 - Long questions (_I have done X and Y and it won't do Z wtf_) - [Stackoverflow](https://stackoverflow.com/questions/tagged/backpack-for-laravel), using the ```backpack-for-laravel``` tag;
 - Showing off something you've made, asking for opinion on Backpack/Laravel matters - [Reddit](https://www.reddit.com/r/BackpackForLaravel/);

Please keep in mind Backpack offers no official / paid support. Whatever help you receive here, on Gitter, Slack or Stackoverflow is thanks to our awesome _awesome_ community members, who give up some of their time to help their peers. If you want to join our community, just start pitching in. We take pride in being a welcoming bunch.

Thank you!

--
Justin Case
The Backpack Robot

**Comment by jcastroa87** [comment]
Hi.

Due to not activity on many years i will close this discussion, if there is needed feel free to reopen or create a new one.

Cheers.
